data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwardingProcessorDecorator.java

```diff
@@ -67,7 +67,16 @@ public static List<Processor> decorateProcessors(
                     "Peer Forwarder Plugin: %s cannot have empty identification keys." + pluginId);
         }
 
-        final PeerForwarder peerForwarder = peerForwarderProvider.register(pipelineName, pluginId, identificationKeys, pipelineWorkerThreads);
+        long localModeProcessors = processors.stream()
+            .filter(processor -> ((RequiresPeerForwarding)processor).isForLocalProcessingOnly(null))
+            .count();
+        final PeerForwarder peerForwarder;
+        if (localModeProcessors < processors.size()) {
+            peerForwarder = peerForwarderProvider.register(pipelineName, pluginId, identificationKeys, pipelineWorkerThreads);
```

Comment:
I'm not sure this logic belongs here. One thought I had was to put it in the `PeerForwarderProvider` class. But, I think a better approach would be to make `PeerForwarderProvider` an interface. Then we could have two implementations. The current `PeerForwarderProvider` could be renamed to something like `DefaultPeerForwarderProvider`.

Then, this logic could be in a `LocalModePeerForwarderProvider` or some other similar name.

You can then use a decorator pattern to decorate them in the `PeerForwarderAppConfig`.

Since we use an application context, we can easily wire it up.

```
    @Bean(name = "defaultPeerForwarder")
    public DefaultPeerForwarderProvider peerForwarderProvider(final PeerForwarderClientFactory peerForwarderClientFactory,
                                                       final PeerForwarderClient peerForwarderClient,
                                                       final PeerForwarderConfiguration peerForwarderConfiguration,
                                                       @Qualifier("peerForwarderMetrics") final PluginMetrics pluginMetrics) {
        return new DefaultPeerForwarderProvider(peerForwarderClientFactory, peerForwarderClient, peerForwarderConfiguration, pluginMetrics);
    }

    @Bean
    @Primary
    public PeerForwarderProvider peerForwarderProvider(@Named("defaultPeerForwarder") final PeerForwarderProvider peerForwarderProvider) {
        return new LocalModePeerForwarderProvider(peerForwarderProvider);
    }

```

This should help decouple the logic for deciding from the decorator which should just be doing the part of using these peer forwarders.

Relatedly, there seems to be something wrong with this method. It is called in a loop for each processor. And then it decorates all processors. Maybe we should fix this at the same time.

Reply:
>>>PeerForwarderProvider

We should create like centralForwarderProvider for S3 buffering based aggregations.

Reply:
@sharraj I think it is much easier to send to S3 via sink than a S3 version of peer forwarder